apps/web/components/dashboard-components/project-details/custom-llm-parameters/editor.tsx

```diff
@@ -0,0 +1,352 @@
+  selectedProvider?: string | null;
+  selectedModel?: string | null;
+  onEdited?: () => void;
+}
+
+/**
+ * CustomLlmParametersEditor Component
+ *
+ * Manages custom LLM parameters with a nested storage structure:
+ * provider -> model -> parameters
+ *
+ * Architecture Decision:
+ * We store parameters per model rather than per provider because:
+ * - Different models may need different optimizations
+ * - Users can switch between models without losing settings
+ * - Allows fine-tuning for specific use cases per model
+ *
+ * The trade-off is added complexity in this component and ai-sdk-client.ts,
+ * but it provides better UX for power users who need model-specific configs.
+ */
+export function CustomLlmParametersEditor({
+  project,
+  selectedProvider,
+  selectedModel,
+  onEdited,
+}: CustomLlmParametersEditorProps) {
+  const { toast } = useToast();
+  const [isEditing, setIsEditing] = useState(false);
+  const [parameters, setParameters] = useState<ParameterEntry[]>([]);
+  const [activeEditIndex, setActiveEditIndex] = useState<number | null>(null);
+
+  // Memoized values
+  const providerName = useMemo(
+    () => selectedProvider || project?.defaultLlmProviderName || "openai",
+    [selectedProvider, project?.defaultLlmProviderName],
+  );
+
+  const currentProvider = useMemo(
+    () => selectedProvider || project?.defaultLlmProviderName,
+    [selectedProvider, project?.defaultLlmProviderName],
+  );
+
+  const currentModel = useMemo(() => {
+    if (selectedModel) return selectedModel;
+
+    // For OpenAI-compatible providers, use customLlmModelName
+    if (providerName === "openai-compatible") {
+      return project?.customLlmModelName;
+    }
+
+    // For other providers, use defaultLlmModelName
+    return project?.defaultLlmModelName;
+  }, [
+    selectedModel,
+    project?.defaultLlmModelName,
+    project?.customLlmModelName,
+    providerName,
+  ]);
+
+  // Determine if custom parameters are allowed for this provider
+  // Only OpenAI-compatible providers can add custom parameters
+  const allowCustomParameters = useMemo(
+    () => providerName === "openai-compatible",
+    [providerName],
+  );
+
+  /**
+   * Extracts parameters from the nested storage structure (provider -> model -> parameters)
+   * and converts them to the UI format for editing
+   */
+  const extractParameters = useCallback(
+    (
+      customParams: unknown,
+      provider?: string | null,
+      model?: string | null,
+    ): ParameterEntry[] => {
+      if (!customParams || !provider || !model) return [];
+      if (typeof customParams !== "object") return [];
+
+      const providerParams = (customParams as Record<string, unknown>)[
+        provider
+      ];
+      if (!providerParams || typeof providerParams !== "object") return [];
+
+      const modelParams = (providerParams as Record<string, unknown>)[model];
+      if (!modelParams || typeof modelParams !== "object") return [];
```

Comment:
```suggestion
      const modelParams = customParams?.[provider]?.[model] ?? {};
```

You don't have to do all this type checking...

Reply:
ahh yes, updated this, [here](https://github.com/tambo-ai/tambo-cloud/pull/1784/commits/b8ce744b7928451cf7fa3eeee4e5dccdddb3abc9)